File: integration-prototypes/nodejs/kafka-ha/producer.js
const { Kafka } = require("kafkajs");

function buildMessage(id, text) {
  return { id, text };
}

async function sendMessage(
  { topic = "prototype-topic-ha", brokers = "localhost:9092,localhost:9094,localhost:9096", message },
  kafkaFactory = (config) => new Kafka(config),
) {
  const kafka = kafkaFactory({
    clientId: "node-kafka-ha-producer",
    brokers: brokers.split(",").map((entry) => entry.trim()),
  });

  const producer = kafka.producer();
  await producer.connect();
  await producer.send({ topic, messages: [{ value: JSON.stringify(message) }] });
  await producer.disconnect();
}

async function main() {
  const payload = buildMessage("1", "hello from node kafka ha producer");
  await sendMessage({
    topic: process.env.KAFKA_TOPIC || "prototype-topic-ha",
    brokers: process.env.KAFKA_BOOTSTRAP_SERVERS || "localhost:9092,localhost:9094,localhost:9096",
    message: payload,
  });
  // eslint-disable-next-line no-console
  console.log("message sent", payload);
}

module.exports = { buildMessage, sendMessage };

if (require.main === module) {
  main().catch((error) => {
    // eslint-disable-next-line no-console
    console.error(error);
    process.exit(1);
  });
}

File: integration-prototypes/python/kafka-ha/producer.py
"""Kafka HA producer prototype."""

from __future__ import annotations

import json
import os
from typing import Any, Callable

try:
    from kafka import KafkaProducer
except ImportError:  # pragma: no cover
    KafkaProducer = None  # type: ignore[assignment]


def build_message(message_id: str, text: str) -> dict[str, str]:
    return {"id": message_id, "text": text}


def send_message(
    topic: str,
    message: dict[str, Any],
    bootstrap_servers: str | None = None,
    producer_factory: Callable[..., Any] | None = None,
) -> None:
    servers = bootstrap_servers or os.getenv("KAFKA_BOOTSTRAP_SERVERS", "localhost:9092,localhost:9094,localhost:9096")
    factory = producer_factory or KafkaProducer
    if factory is None:
        raise RuntimeError("kafka-python is not installed")

    producer = factory(
        bootstrap_servers=servers,
        value_serializer=lambda value: json.dumps(value).encode("utf-8"),
    )
    producer.send(topic, message).get(timeout=10)
    producer.flush()
    producer.close()


def main() -> None:
    topic = os.getenv("KAFKA_TOPIC", "prototype-topic-ha")
    payload = build_message("1", "hello from python kafka ha producer")
    send_message(topic=topic, message=payload)
    print(f"sent message to {topic}: {payload}")


if __name__ == "__main__":
    main()
